s2a/src/main/java/io/grpc/s2a/internal/handshaker/S2AProtocolNegotiatorFactory.java

```diff
@@ -233,9 +254,12 @@ public void onSuccess(SslContext sslContext) {
                           SharedResourcePool.forResource(GrpcUtil.SHARED_CHANNEL_EXECUTOR))
                       .newHandler(grpcHandler);
 
-              // Remove the bufferReads handler and delegate the rest of the handshake to the TLS
-              // handler.
-              ctx.pipeline().addAfter(ctx.name(), /* name= */ null, handler);
+              // Delegate the rest of the handshake to the TLS handler. and remove the 
+              // bufferReads handler.
+              ctx.pipeline().addAfter(ctx.name(), "tlsHandler", handler);
+              ctx.pipeline().addAfter("tlsHandler", "cleanupHandler",
+                  new S2AStubCleanupNegotiationHandler(handler,
```

Comment:
This is adding the same handler twice. Looks like we can instead pass the cleanup handler to `tls(...).newHandler(cleanupHandler)` and have the tls handler add it to the pipeline for us. The `grpcHandler` should be passed to this cleanup handler.

I'm concerned that no tests failed as the code is now. I expect `s2aStub.close();` is not being called. I don't know how hard it'd be to detect if a leak happens in the test (e.g., by checking that the server stream still exists), but it is worth consideration (but not required).

Reply:
If stub.close() is called twice, the second one is a no-op.  I believe that the current code runs tls->cleanup->tls which shouldn't exhibit a problem.

Reply:
Wouldn't s2aStub.close() be called once, but then the second time the tls handler was run it would reopen it and that instance wouldn't be closed?

Reply:
Thanks for catching this adding of the tls handler twice. Looking at `tls(...).newHandler(...)` it expects a GrpcHttp2ConnectionHandler, so I don't think we can pass in the custom `S2AStubCleanupNegotiationHandler` to it. I added the `S2AStubCleanupNegotiationHandler` via`addLast`, WDYT?

Reply:
Hmm... That's a problem. The current code doesn't seem to do anything. (It seems we really need a test to check the lifecycle.)

I'm not immediately seeing any great ways of dealing with this. I'd be interested in how the lifecycle is managed in other languages.

Reply:
In golang S2A client, we close the stream to S2A after the handshake: https://github.com/google/s2a-go/blob/main/internal/v2/s2av2.go#L152

Reply:
Because the handshake is just a method, that does make it easy to know when it is done. The equivalent in Java is `SSLEngine.getHandshakeStatus() == FINISHED`, but that's not at all easy to detect.

The easiest solution seems to be to pass a (optional) `Runnable` ("handshakeCompleteRunnable"?) to `InternalProtocolNegotiators.tls()` and plumb that through `ClientTlsProtocolNegotiator` and `ClientTlsHandler`. Then in `ClientTlsHandler.propagateTlsComplete()` we call the Runnable. We also need to handle the case where the connection is closed before the handshake finishes; seems calling that same runnable when `userEventTriggered0()` calls `fireExceptionCaught()` would be the easier way to manage that.

Reply:
Done in https://github.com/grpc/grpc-java/pull/11600/commits/1698b54622602a30f9c3654ec8841e8619f7191d. Thanks for pointing out that corner case where the connection gets closed before the end of the handshaker. I added a new integration test to verify stub gets closed. Confirmed (manally) that not setting the `handshakeCompleteRunnable` field results in stub not getting closed.